content/post/codesent.md

```diff
@@ -0,0 +1,396 @@
+It was created by David Crawshaw, the author of `pkg/plugin`. Neugram is still in its early ages
+and has youth related issues, but I really like it. Who knows, maybe one day we will train Tensorflow
+models using neugram - I anticipate this.
+
+The following Go code runs a Tensorflow graph:
+
+```go
+modelBytes, _ := assets.Asset("model.pb")
+graph := tf.NewGraph()
+graph.Import(modelBytes, "")
+input := graph.Operation("input").Output(0)
+output := graph.Operation("output").Output(0)
+sess, _ := tf.NewSession(graph, &tf.SessionOptions{})
+feed, _ := tf.NewTensor([][]float32{})
+result, _ := sess.Run(
+    map[tf.Output]*tf.Tensor{input: input1},
+    []tf.Output{output}, nil)
+```
+
+We embed the model using [go-bindata](https://github.com/jteeuwen/go-bindata), import the graph and
+discover input and output tensors. The new session is launched and we execute the resolved dependency
+tree to reach the output given the input. Sessions are required to allocate the hardware resources
+and manage the environment. Technically, they are bridges to the underlying Tensorflow runtime
+written in C++ as a standalone library, in case with Go this bridge is implemented with CGo.
+
+### BiDiSentiment
+
+{{% caption src="/post/codesent/gophers.png" %}}
+vmarkovtsev/BiDiSentiment logo.
+{{% /caption %}}
+
+What's left is to reveal the comment sentiment classification model. I named it "BiDiSentiment" and
+this is it's logo. BiDiSentiment is a general-purpose sentiment model, it is not specifically designed
+for comments.
+
+{{% caption src="/post/codesent/arch_black.png" %}}
+BiDiSentiment architecture.
+{{% /caption %}}
+
+The model is by far not the state-of-the-art but it works reasonably well. It consists of two
+stacked recurrent branches, each branch has two LSTM layers. Both read text byte by byte, the first
+from left to right and the second in the opposite direction. Once all the text is read, their
+outputs are concatenated and the dense layer put on top. The result is two numbers: the estimated
+probabilities of the negative and the positive sentiment. Obviously, their sum is 1.
+
+Since the recurrent branches are independent, we can pin them to different GPUs which speeds up
+the training by 50%.
+
+{{% caption src="/post/codesent/arch_device_black.png" %}}
+BiDiSentiment training device allocation.
+{{% /caption %}}
+
+The biggest labelled dataset that I found was 1.5 million tweets so I trained on it. As a bonus,
+the text sizes are bound to 140 (now 280) characters which perfectly suits the maximum recurrent memory
+depth. Here is the plot of the accuracy (number of correctly predicted sentiments divided by the overall
+size) through time.
+
+{{% caption src="/post/codesent/validation_black.png" %}}
+BiDiSentiment accuracy on validation.
+{{% /caption %}}
+
+We split the data into two parts: the part on which we train and the part on which we evaluate aka
```

Comment:
I think you should use the well established terms here: training and validation datasets.

We split our full dataset into two parts:

- the training dataset contains X% of randomly selected samples, and it's used to train our model,
- the validation dataset contains the remaining %X and it's used to evaluate the model's accuracy.